### PR TITLE
[SPARK-43929][SPARK-44073][SQL][PYTHON][CONNECT][FOLLOWUP] Add extract, date_part, datepart to Scala, Python and Connect API

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR follows up https://github.com/apache/spark/pull/41636 and https://github.com/apache/spark/pull/41651 and add extract, date_part, datepart to Scala, Python and Connect API.

### Why are the changes needed?
Add extract, date_part, datepart to Scala, Python and Connect API

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

### CODE REVIEW

**Review by zhengruifeng — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -4557,6 +4557,47 @@ object functions {
...
+   * Extracts a part of the date/timestamp or interval source.
+   *
+   * @param field selects which part of the source should be extracted.
+   * @param source a date/timestamp or interval column from where `field` should be extracted.
+   * @return a part of the date/timestamp or interval source
+   * @group datetime_funcs
+   * @since 3.5.0
+   */
+  def extract(field: Column, source: Column): Column = withExpr {
+    UnresolvedFunction("extract", Seq(field, source).map(_.expr), isDistinct = false)
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4570]: I just notice that we directly use `UnresolvedFunction` before, e.g. `floor` / `lpad` / `ceil`

would you mind also change `endswith`, `startswith`, `contains` in this PR

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4570]: to narrow the related tickets, let me send the followup

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: @zhengruifeng Sorry, I saw the comment too late.

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: no worries, I send it in another pr

**Comment by beliefer** [comment]
@zhengruifeng Thank you!

**Pull request closed without merging**